### PR TITLE
Support all value types

### PR DESCRIPTION
**Pull request by otoolep**

### CODE REVIEW

**Comment by otoolep** [comment]
WIP.

File: shard.go [hunk 1]
```diff
@@ -180,14 +179,31 @@ func marshalValues(values map[uint8]interface{}) []byte {
 		}
 
 		// Encode value after field id.
-		// TODO: Support non-float types.
 		switch v := v.(type) {
 		case float64:
+			buf = make([]byte, 9)
 			binary.BigEndian.PutUint64(buf[1:9], math.Float64bits(v))
+		case bool:
+			buf = make([]byte, 2)
```
Comment (otoolep) [inline shard.go]: @benbjohnson -- from implementing this function, it seems to me we need to code more information in a (field ID, value) pair -- we must also encode the type. Can you confirm? `unmarshalValues` just assumes that it's a float type in the database, but that will no longer be true with these changes.

If I am correct, a single byte after the field ID would be sufficient.

Comment (pauldix) [inline shard.go]: You shouldn't have to encode the type as that's kept in the metastore. Field id, followed by value (if a fixed size type) or in the case of bytes and strings, field id followed by length, followed by value.

Comment (otoolep) [inline shard.go]: So why is this code casting it to a float before sending it up? Is that redundant? It did occur to me that the info is in the metastore, and we could decode at a higher level.

Comment (otoolep) [inline shard.go]: To be clear -- the casting is happening in `unmarshalValues`.

Comment (otoolep) [inline shard.go]: It seems like I could preform the type conversion using the metastore around here:

https://github.com/influxdb/influxdb/blob/master/server.go#L1839

so the low-level conversion to float (and support for other types at this low-level as implied by the comments) may be unnecessary.

Comment (otoolep) [inline shard.go]: Ah -- unless your distinction is just between numeric (floats) with a fixed length and variable length like strings? I interpreted the comments to mean we needed a breakdown by all supported types at this level.

Comment (pauldix) [inline shard.go]: I think the `ReadSeries` function is actually cruft left over from stuff that was in there for debugging purposes. The real way to read raw series data is through the query engine, which now actually works.

There's some other spot in the code where you can get a value based on the field ID. At some point someone needs to cast this to the correct data type. That should be done in the query engine before it gets yielded to the mappers.

Comment (otoolep) [inline shard.go]: Ah -- OK. I do recall @benbjohnson remarking that `ReadSeries` was debug. If casting strictly happens at the higher level, yeah, then at this low-level it really is just a bunch of bytes.

Let me take another pass.

**Comment by otoolep** [comment]
Just for the record, `unmarshalValues` is also part of the production shard, and is called while iterating over the shard (I had assumed after the discussion above that is was actually just by `ReadSeries`). So right now all type "conversion" is being done at the low-level, and no reference is made to the Metatstore when performing the conversion, that I can see.

**Comment by otoolep** [comment]
Obviously using a single source of truth for field types would be best, of course.

**Comment by otoolep** [comment]
@pauldix @benbjohnson -- please review these latest changes and let me know what you think. This is still WIP, but I think it's close. I'd like some feedback at this point. Points to note:

* I basically just lifted the "marshal" and "unmarshal" code from `shard.go`. There are some potential optimizations there -- and we should do them -- but wanted to get this working first.
* I just added a new `interface` to the cursor, such that it expects a field decoder. I think this is fine, but for now I just slammed the `Measurement` object in there. Doesn't feel great.
* @benbjohnson suggested a dedicated field codec, with a static of set of fields, presumably to avoid races. But we seem to have the same race with `mapValues` and there is no locking there. Is this code special, or we just haven't fixed up `mapValues` yet?
* All integers are still stored internally as floats. 
* While the build is green, I haven't actually added any tests for booleans and strings yet.

**Comment by otoolep** [comment]
I just noticed there is one more place that needs change:

https://github.com/influxdb/influxdb/blob/master/server.go#L1682

**Comment by pauldix** [comment]
We need to also support raw bytes as a type.

On the mapValues part of things, almost all the aggregate functions look at only one field. It would be nice to have a function that takes this field ID and returns the raw value without marshaling it into a map. We should be able to just cast the raw bytes from the DB into the type.

This would save us a ton when aggregating a through a bunch of data by not allocating a bunch of maps.

**Comment by otoolep** [comment]
How would a user insert raw bytes?

**Comment by benbjohnson** [comment]
Materializing into a map was just to get it working. It definitely need to get optimized. As for the `mapValues()`, the lack of locking is a bug. Queries can potentially run for a while so we need to be able to copy the field metadata to the codec to be used during query execution.

**Comment by pauldix** [comment]
I'm looking at some other stuff on the query side in terms of raw data queries. I think it would be useful to have the iterators yield the raw bytes to the map functions. Then have the iterator have a method that will pull a field value from those bytes.

This will be useful for queries that need to access multiple fields in the map portion of the query.

**Comment by otoolep** [comment]
Field codec split into its own type, which allows the locking to be decoupled from the server lock. I still need to add support for other types, and have questions about interactions with the shard iterators.

**Comment by otoolep** [comment]
Unit tests added for string and boolean types.

**Comment by pauldix** [comment]
LGTM. Drop it like it's hot. Where's the Snoop Dogg emoji when you need it?

**Comment by otoolep** [comment]
Thanks for the review @pauldix -- merging now.

**Pull request merged**